huaweicloud/services/iotda/resource_huaweicloud_iotda_upgrade_package.go

```diff
@@ -0,0 +1,252 @@
+	if raw[0] == nil {
+		return nil
+	}
+
+	obsLocationParam := buildObsLocationParam(raw[0].(map[string]interface{})["obs_location"])
+	fileLocationParam := model.FileLocation{
+		ObsLocation: obsLocationParam,
+	}
+
+	return &fileLocationParam
+}
+
+func buildUpgradePackageCCreateParams(d *schema.ResourceData) *model.CreateOtaPackageRequest {
+	req := model.CreateOtaPackageRequest{
+		Body: &model.CreateOtaPackage{
+			AppId:                 d.Get("space_id").(string),
+			PackageType:           d.Get("type").(string),
+			ProductId:             d.Get("product_id").(string),
+			Version:               d.Get("version").(string),
+			FileLocation:          buildFileLocationParam(d.Get("file_location").([]interface{})),
+			SupportSourceVersions: utils.ExpandToStringListPointer(d.Get("support_source_versions").([]interface{})),
+			Description:           utils.StringIgnoreEmpty(d.Get("description").(string)),
+			CustomInfo:            utils.StringIgnoreEmpty(d.Get("custom_info").(string)),
+		},
+	}
+
+	return &req
+}
+
+func resourceUpgradePackageCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	createOpts := buildUpgradePackageCCreateParams(d)
+	resp, err := client.CreateOtaPackage(createOpts)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA OTA upgrade package: %s", err)
+	}
+
+	if resp == nil || resp.PackageId == nil {
+		return diag.Errorf("error creating IoTDA OTA upgrade package: ID is not found in API response")
+	}
+
+	d.SetId(*resp.PackageId)
+
+	return resourceUpgradePackageRead(ctx, d, meta)
+}
+
+func resourceUpgradePackageRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	getOpts := &model.ShowOtaPackageRequest{
+		PackageId: d.Id(),
+	}
+	resp, err := client.ShowOtaPackage(getOpts)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error querying IoTDA OTA upgrade package")
+	}
+
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("space_id", resp.AppId),
+		d.Set("type", resp.PackageType),
+		d.Set("product_id", resp.ProductId),
+		d.Set("version", resp.Version),
+		d.Set("file_location", flattenFileLocation(resp.FileLocation)),
+		d.Set("support_source_versions", resp.SupportSourceVersions),
+		d.Set("description", resp.Description),
+		d.Set("custom_info", resp.CustomInfo),
+		d.Set("created_at", resp.CreateTime),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func flattenFileLocation(resp *model.FileLocation) []interface{} {
+	obsLocation := flattenObsLocation(resp.ObsLocation)
```

Comment:
For pointer types, perform null judgment in advance.
Same to function `flattenObsLocation`.

Reply:
fixed.